app/src/main/java/org/thoughtcrime/securesms/components/WaveFormSeekBarView.java

```diff
@@ -0,0 +1,164 @@
+    super(context);
+    init();
+  }
+
+  public WaveFormSeekBarView(Context context, @Nullable AttributeSet attrs) {
+    super(context, attrs);
+    init();
+  }
+
+  public WaveFormSeekBarView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
+    super(context, attrs, defStyleAttr);
+    init();
+  }
+
+  private void init() {
+    setWillNotDraw(false);
+    
+    paint.setStrokeCap(Paint.Cap.ROUND);
+    paint.setAntiAlias(true);
+
+    progressDrawable = super.getProgressDrawable();
+
+    if (isInEditMode()) {
+      setWaveData(sinusoidalExampleData());
+      dataSetTime = 0;
+    }
+
+    barWidth = getResources().getDimensionPixelSize(R.dimen.wave_form_bar_width);
+  }
+
+  public void setColors(@ColorInt int playedBarColor, @ColorInt int unplayedBarColor, @ColorInt int thumbTint) {
+    this.playedBarColor   = playedBarColor;
+    this.unplayedBarColor = unplayedBarColor;
+
+    getThumb().setColorFilter(thumbTint, PorterDuff.Mode.SRC_IN);
+
+    invalidate();
+  }
+
+  @Override
+  public void setProgressDrawable(Drawable progressDrawable) {
+    this.progressDrawable = progressDrawable;
+    if (!waveMode) {
+      super.setProgressDrawable(progressDrawable);
+    }
+  }
+
+  @Override
+  public Drawable getProgressDrawable() {
+    return progressDrawable;
+  }
+
+  public void setWaveData(@NonNull float[] data) {
+    if (!Arrays.equals(data, this.data)) {
+      this.data        = data;
+      this.dataSetTime = System.currentTimeMillis();
+    }
+    setWaveMode(data.length > 0);
+  }
+
+  public void setWaveMode(boolean waveMode) {
+    this.waveMode = waveMode;
+    super.setProgressDrawable(this.waveMode ? null : progressDrawable);
+    invalidate();
+  }
+
+  @Override
+  protected void onDraw(Canvas canvas) {
+    if (waveMode) {
+      drawWave(canvas);
+    }
+    super.onDraw(canvas);
+  }
+
+  private void drawWave(Canvas canvas) {
+    paint.setStrokeWidth(barWidth);
+
+    int   usableHeight = getHeight() - getPaddingTop() - getPaddingBottom();
+    int   usableWidth  = getWidth() - getPaddingLeft() - getPaddingRight();
+    float midpoint     = usableHeight / 2f;
+    float maxHeight    = usableHeight / 2f - barWidth;
+    float barGap       = (usableWidth - data.length * barWidth) / (float) (data.length - 1);
+
+    boolean hasMoreFrames = false;
+
+    canvas.save();
+    canvas.translate(getPaddingLeft(), getPaddingTop());
+
+    if (getLayoutDirection() == LAYOUT_DIRECTION_RTL) {
+      canvas.scale(-1, 1, usableWidth / 2f, usableHeight / 2f);
+    }
+
+    for (int bar = 0; bar < data.length; bar++) {
+      float x        = bar * (barWidth + barGap) + barWidth / 2f;
+      float y        = data[bar] * maxHeight;
+      float progress = x / usableWidth;
+
+      paint.setColor(progress * getMax() < getProgress() ? playedBarColor : unplayedBarColor);
+
+      long  time             = System.currentTimeMillis() - bar * ANIM_BAR_OFF_SET_DURATION - dataSetTime;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).